### PR TITLE
When using `process::Command` on Windows, environment variable names must be case-preserving but case-insensitive

### LINKED ISSUE #85242: Windows std::env::vars() uppercases environment variables
I tried this code:

```rust
    for v in std::env::vars() {
        println!("{:?}", v);
    }
```

I expected to see this happen: a case-sensitive list of environment variables and values, matching the output of `Environment.GetEnvironmentVariables` from C# and .NET and the output of `Get-ChildItem Env:` from PowerShell.

It seems we should align the `std::env::vars()` output to match the modern APIs for Windows that Microsoft provides.

Instead, this happened: `std::env::vars()` returns all environment variable names in all uppercase.

Rust via `std::env::vars()`:
```
SYSTEMDRIVE = C:
<snip>
```

C# via `Environment.GetEnvironmentVariables`:
```
SystemDrive = C:
<snip>
```

### Meta

`rustc --version --verbose`:
```
>rustc --version --verbose
rustc 1.51.0-nightly (04caa632d 2021-01-30)
binary: rustc
commit-hash: 04caa632dd10c2bf64b69524c7f9c4c30a436877
commit-date: 2021-01-30
host: x86_64-pc-windows-msvc
release: 1.51.0-nightly
LLVM version: 11.0.1
```

</p>
</details>

### PR DESCRIPTION
**Pull request by ChrisDenton**
When using `Command` to set the environment variables, the key should be compared as uppercase Unicode but when set it should preserve the original case.

Fixes #85242

### CODE REVIEW

**Comment by ChrisDenton** [comment]
I've moved the test to the right place. Sorry about that.

File: library/std/src/sys/windows/process/tests.rs [hunk 1]
```diff
@@ -41,3 +42,28 @@ fn test_make_command_line() {
...
+// See: #85242
+#[test]
+fn windows_env_unicode_case() {
+    let lowercase = "ä";
+    let uppercase = "Ä";
+
+    let mut cmd = Command::new("cmd");
+    cmd.env(lowercase, "1").env(uppercase, "2");
+
+    // At this ponit "ä" should equal "2" despite setting it to "1"
```
Comment (CDirkx) [inline library/std/src/sys/windows/process/tests.rs:56]:
```suggestion
    // At this point "ä" should equal "2" despite setting it to "1"
```

File: library/std/src/sys/windows/process/tests.rs [hunk 2]
```diff
@@ -41,3 +42,28 @@ fn test_make_command_line() {
         "\"\u{03c0}\u{042f}\u{97f3}\u{00e6}\u{221e}\""
     );
 }
+
+// On Windows, environmet args are case preserving but comparisions are case-insensitive.
```
Comment (CDirkx) [inline library/std/src/sys/windows/process/tests.rs:46]:
```suggestion
// On Windows, environment args are case preserving but comparisons are case-insensitive.
```

File: library/std/src/sys/windows/process.rs [hunk 3]
```diff
@@ -31,13 +32,66 @@ use libc::{c_void, EXIT_FAILURE, EXIT_SUCCESS};
...
 pub struct EnvKey(OsString);
 
+// Windows environment variables are case preserving but comparisions are case
+// insensitive. So we need to convert to uppercase. This converts UTF-16 to
+// UTF-32 code units, converting the case along the way. Unpaired surrogates are
+// left as they are.
+fn iter_uppercase<T: Iterator<Item = u16>>(iter: T) -> impl Iterator<Item = u32> {
+    char::decode_utf16(iter).flat_map(|cp| match cp {
+        Err(e) => CodeUnits::Unit(Some(e.unpaired_surrogate())),
+        Ok(cp) => CodeUnits::Iter(cp.to_uppercase()),
```
Comment (CDirkx) [inline library/std/src/sys/windows/process.rs:46]: I checked and `to_uppercase` is not always correct: `'ß'.to_uppercase() == "SS"`, but the following fails
```rust
set_var("ß", "something");
var("SS").unwrap() // panic: Err<NotPresent>
```

I can't find anything about in official documentation, but from tests I did I suspect Windows uses simple case folding (for context see [Unicode Technical Report #21 Case Mappings Section 1.3](https://www.unicode.org/reports/tr21/tr21-5.html), simple case folding is the unicode case folding algorithm, but restricted to code points that don't expand under upper/lower casing like `ß`). I also checked if any normalization or locale-dependent behavior was happening, but that doesn't appear to be the case: `é` and `é` (e + acute accent combining character) are different keys and nothing changed around `I` on the Turkish locale (https://en.wikipedia.org/wiki/Dotted_and_dotless_I).

Not sure what the best way to use case folding in std would be (if that is even what Windows is indeed using). There are crates like [unicode-rs/rust-caseless](https://github.com/unicode-rs/rust-caseless) implementing case folding, but I couldn't find any that also implements the simple variant.

Comment (ChrisDenton) [reply library/std/src/sys/windows/process.rs]: Hm yeah, that is a problem. I think the only way to resolve this properly is to use the OS's own function for this. So in the next commit I've replaced my implementation with a Windows API call and expanded the testing slightly.

Comment (CDirkx) [reply library/std/src/sys/windows/process.rs]: I looked through the docs some more: 

File paths and environment variables should be compared using `StringComparison.OrdinalIgnoreCase` [1]. Comparisons made using `OrdinalIgnoreCase` are behaviorally the composition of two calls: calling `ToUpperInvariant` on both string arguments, and doing a comparison using `Ordinal` [2]. A comparison using `Ordinal` is just a comparison based on the numerical value of each utf16 code unit [3]. 

So the original approach was conceptually correct, but Windows apparently handles converting to uppercase differently than rust (I haven't yet found anywhere how `ToUpperInvariant` is actually implemented 🙁). I agree with the approach to instead just use the Windows API. 

(As an aside, with the relevant Windows API added a future possibility for `std` would be to expose case insensitive functionality with a `std::os` extension trait, for example for correctly comparing paths case insensitively.)

[1] https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#choosing-a-stringcomparison-member-for-your-method-call
[2] https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#stringtoupper-and-stringtolower
[3] https://docs.microsoft.com/en-us/dotnet/api/system.stringcomparison?view=net-5.0#System_StringComparison_Ordinal

**Comment by Mark-Simulacrum** [comment]
@rustbot ping windows

It sounds like there's some differences between Rust's case-conversion and the one used by Windows - see this comment https://github.com/rust-lang/rust/pull/85270#discussion_r632806641 -- I'm wondering if some windows folks can weigh in on the right thing for us to do here for comparison.

I think it sounds like @ChrisDenton this PR is actually solving two different problems right now, right? We're currently exposing the case-converted (i.e., to_ascii_uppercase) environment variable names through std's APIs, which is loosely the decoupling intended by the first commit here. Then, the second question is whether to_ascii_uppercase is actually the right function to call (sounds like no), and what that function should be.

**Comment by ChrisDenton** [comment]
To sum up where we are at now, there were two problems with the original code, before this PR:

1. It ASCII uppercased environment variable keys. It should have left them as-is.
2. The uppercasing was (possibly) done so that looking up environment variables by key would be case-insensitive. The problem with this is that Windows uses a variant of Unicode casing rules to do key comparisons, not just ASCII uppercasing.

My first attempt to fix this was to remove the original uppercasing and use Rust's Unicode functions to compare (and only to compare) the strings. As [CDirkx has shown](https://github.com/rust-lang/rust/pull/85270#discussion_r632655845), this was wrong because Windows uses what it calls "ordinal" comparisons in this context. This ordinal comparison is documented as testing for "binary equality, not linguistic equality".

So now I'm using the Windows API function [`CompareStringOrdinal`](https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-comparestringordinal) to do the comparison, which appears to do what the OS expects.

**Comment by ChrisDenton** [comment]
Ok so I investigated this further and I'm now even more confident that using the system APIs is the right one. Elsewhere I have figured out the [case-folding algorithm](https://github.com/ChrisDenton/Windows-case) for my system, but it's odd and can vary a bit depending on the OS version. Therefore asking the OS to compare strings is, I think, the best approach.

@CDirkx should I squish my commits to make my changes clearer?

File: library/std/src/sys/windows/c.rs [hunk 4]
```diff
@@ -1070,6 +1074,14 @@ extern "system" {
     pub fn ReleaseSRWLockShared(SRWLock: PSRWLOCK);
     pub fn TryAcquireSRWLockExclusive(SRWLock: PSRWLOCK) -> BOOLEAN;
     pub fn TryAcquireSRWLockShared(SRWLock: PSRWLOCK) -> BOOLEAN;
+
+    pub fn CompareStringOrdinal(
+        lpString1: LPCWSTR,
+        cchCount1: c_int,
+        lpString2: LPCWSTR,
+        cchCount2: c_int,
+        bIgnoreCase: i32,
```
Comment (CDirkx) [inline library/std/src/sys/windows/c.rs:1083]:
```suggestion
        bIgnoreCase: BOOL,
```

Small nit: I would use the type alias `BOOL` to match the signature in the [docs](https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-comparestringordinal) and make it clear that `c::TRUE` and `c::FALSE` are the only allowed options.

File: library/std/src/sys/windows/process.rs [hunk 5]
```diff
@@ -31,32 +32,72 @@ use libc::{c_void, EXIT_FAILURE, EXIT_SUCCESS};
...
+        }
+    }
+}
 
 impl From<OsString> for EnvKey {
-    fn from(mut k: OsString) -> Self {
-        k.make_ascii_uppercase();
-        EnvKey(k)
+    fn from(k: OsString) -> Self {
+        EnvKey { utf16: k.encode_wide().collect(), os_string: k }
```
Comment (CDirkx) [inline library/std/src/sys/windows/process.rs:82]: There is a helper function`sys::windows::to_u16s`. It might be possible to use that instead and remove the `ensure_no_nuls` in this module.
https://github.com/rust-lang/rust/blob/3bcaeb0bf9e1c29d18abc32928fd2f23d1bed0bd/library/std/src/sys/windows/mod.rs#L139-L152

Comment (ChrisDenton) [reply library/std/src/sys/windows/process.rs:82]: My thinking was that the `From` implementation does not allow returning an error whereas `make_envp` does. So I'd have to turn it into a panic in place of the error?

File: library/std/src/sys/windows/process.rs [hunk 6]
```diff
@@ -31,32 +32,72 @@ use libc::{c_void, EXIT_FAILURE, EXIT_SUCCESS};
 // Command
 ////////////////////////////////////////////////////////////////////////////////
 
-#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Clone, Debug, Eq)]
 #[doc(hidden)]
-pub struct EnvKey(OsString);
+pub struct EnvKey {
+    os_string: OsString,
+    utf16: Vec<u16>,
```
Comment (CDirkx) [inline library/std/src/sys/windows/process.rs:39]: It is unfortunate that `OsString` on Windows is backed by an WTF8 encoded `Vec<u8>`, instead of an `Vec<u16>`. Normally in Windows API calls the osstring is only converted to an u16 buffer right before the call (see almost every function in `sys::windows::fs`). However converting and having to allocate for every `==` would be expensive, so storing the utf16 encoding seems like a reasonable solution. Could you add a comment to the field explaining that this is a workaround?

File: library/std/src/sys/windows/process.rs (hunk 5 above)
Comment (CDirkx) [reply library/std/src/sys/windows/process.rs:82]: Mmm yeah that's true. Then I would leave it as is, refactoring things to reduce duplication can always be done in a later as it is not really related to this PR.

**Comment by CDirkx** [comment]
I was going to comment that this behavior should be documented somewhere on the `process::Command` docs, but it already is:

> Note that environment variable names are case-insensitive (but case-preserving) on Windows, and case-sensitive on all other platforms.

(https://doc.rust-lang.org/std/process/struct.Command.html#method.env)

**Comment by CDirkx** [comment]
PR looks good to me, @ChrisDenton could you squash the commits?

**Comment by CDirkx** [comment]
@Mark-Simulacrum this looks ready for review (although I'm not from the Windows group). This PR addresses two problems:
 - env keys on Windows are case-insensitive, but case-preserving. The original implementation stored the `make_ascii_uppercase`d key, making the derived `Ord` implementation effectively case-insensitive (although not completely correct, see the second point), but not preserving case. This PR stores the key as given, instead adding a manual case-insensitive `Ord` implementation, thus preserving case. 
 - The way keys are compared is changed because just uppercasing and comparing is not sufficient, Windows has complex case-folding rules, based on an older version of Unicode case folding, but with additional custom mappings. Instead of re-implementing this functionality in `std` and possibly getting it wrong, the Windows API `CompareStringOrdinal` with `bIgnoreCase = TRUE` is used. This does have some costs however, `OsString` is a `Vec<u8>` and Windows API calls expect `*const u16`. To avoid having to convert (which requires an allocation) every time the `BTreeMap` might compare keys, both the original `OsString` and the converted UTF16-encoded `Vec<u16>` are stored in the key, resulting in a bit more memory use. I don't expect this to be an issue however.

So in conclusion, this PR makes sure that the code actually matches what is already in the documentation, implementing case-preserving and correct case-insensitivity:
> Note that environment variable names are case-insensitive (but case-preserving) on Windows, and case-sensitive on all other platforms.

(https://doc.rust-lang.org/std/process/struct.Command.html#method.env)

**Comment by Mark-Simulacrum** [comment]
r=me on the implementation. This is technically going to break any user expecting the to_uppercase behavior on Windows, right? e.g., previously env("Cargo_incremental", "1") would set CARGO_INCREMENTAL, but now won't. Seems like an unlikely situation but we'll likely want to FCP and relnotes if we land this.

**Comment by CDirkx** [comment]
I don't think there would really be such an expectation. The docs already stated that on windows env vars are case-preserving and in general that is what Windows users would expect. I would consider the current implementation a bug, as the to_uppercase doesn't preserve case and does thus not match the documentation.

There is however indeed the small change of breaking changes; the OS will handle "CARGO_INCREMENTAL" and "Cargo_incremental" the same but applications might not.

**Comment by Mark-Simulacrum** [comment]
I agree that the current implementation is a bug. That doesn't necessarily mean we can 'just' break things for users regardless, though. @rust-lang/libs, would someone care to weigh in here? I don't know if an FCP is in order, but as this is technically a breaking change it seems potentially appropriate.

**Comment by yaahc** [comment]
> I agree that the current implementation is a bug. That doesn't necessarily mean we can 'just' break things for users regardless, though. @rust-lang/libs, would someone care to weigh in here?

> I don't know if an FCP is in order, but as this is technically a breaking change it seems potentially appropriate.

IMO we should FCP this, and we should probably also formalize our policy on this sort of runtime behavior breakage since this [come up recently](https://github.com/rust-lang/project-error-handling/issues/44#issuecomment-845398028). IMO we should treat this as a more serious form of breakage than ones detectable at compile time due to the silent nature of the breakage, so FCP is super warranted.

**Comment by CDirkx** [comment]
Yes I believe the only way this could lead to problems is if the user is getting the environment variable in the child process via a way that doesn't take the case-insensitivity into account (which can happen in other languages, like with [getenv](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/getenv-wgetenv?view=msvc-160), or in Rust by directly comparing against values returned by `std::env::vars`) AND the user either didn't take this into account or was explicitly relying on the (undocumented) uppercasing, so setting "var" while the child process expects "VAR".

How likely is that? I don't really know. I can kind of see it happening that someone sets "Path" (reasonably common spelling on Windows) and calls a C program that only checks exactly "PATH". That works with the current code, but would break with this PR. Not sure how we would test this.

**Comment by ChrisDenton** [comment]
> How likely is that? I don't really know. I can kind of see it happening that someone sets "Path" (reasonably common spelling on Windows) and calls a C program that only checks exactly "PATH". That works with the current code, but would break with this PR. Not sure how we would test this.

The C program would have to be manually parsing the environment string instead of using `GetEnvironmentVariable` (as well as doing a case sensitive comparison). This would already be broken for keys that canonically have lower case ASCII characters because these are currently impossible to set using `std`. For example, `windir`, `SystemDrive`, `ProgramFiles`, etc. But I'd agree there is potential (however unlikely) for currently working code to break after this PR.

I would also reiterate that the [current documentation](https://doc.rust-lang.org/std/process/struct.Command.html#method.env) is written as if this PR were merged. Perhaps at some point Rust did work correctly?

**Comment by yaahc** [comment]
Yea, I'm not really sure how we'd investigate this either. If they have test cases that test that behavior a crater run might catch it, but that seems unlikely. The only other option I can think of is to lookup all crates that use `.env` on a `std::process::Command` and audit them to look for likely suspects, but this seems unfairly labor intensive to ask for.

**Comment by ChrisDenton** [comment]
To sum up, the current behaviour is flawed because:

1. It does not match the OS behaviour.
2. It does not match the documented behaviour.
3. It's inconsistent. For ASCII it unconditionally upper cases characters a-z. For non-ASCII it preserves the case but does a bit-for-bit compare. Both behaviours are wrong but in different ways.

Realistically, for this PR to break current code this must happen:

1. The environment variable is canonically ASCII upper case.
2. The Rust program uses `Command` and sets the environment variable using one or more ASCII lower case characters.
3. A child process both manually parses the environment block and does case-sensitive key matching.

Note that the child process would already be broken for any parent process that does not ensure the relevant environment variables are ASCII upper cased. For example, if you run a program using `cargo run` then all environment variables will be ASCII upper cased but if you run the program directly then the environment variables will preserve their natural case. Either way `std::env::var` will work the same as it uses the OS function `GetEnvironmentVariableW`.

In short, I'm not entirely sure how you'd look for likely suspects. I guess if they set a `Command` var with lower case ASCII? But that's only a problem if the child process absolutely requires ASCII upper case.

**Comment by yaahc** [comment]
> To sum up, the current behaviour is flawed because:
> 
>     1. It does not match the OS behaviour.
> 
>     2. It does not match the documented behaviour.
> 
>     3. It's inconsistent. For ASCII it unconditionally upper cases characters a-z. For non-ASCII it preserves the case but does a bit-for-bit compare. Both behaviours are wrong but in different ways.
> 
> 
> Realistically, for this PR to break current code this must happen:
> 
>     1. The environment variable is canonically ASCII upper case.
> 
>     2. The Rust program uses `Command` and sets the environment variable using one or more ASCII lower case characters.
> 
>     3. A child process both manually parses the environment block and does case-sensitive key matching.
> 
> 
> Note that the child process would already be broken for any parent process that does not ensure the relevant environment variables are ASCII upper cased. For example, if you run a program using `cargo run` then all environment variables will be ASCII upper cased but if you run the program directly then the environment variables will preserve their natural case. Either way `std::env::var` will work the same as it uses the OS function `GetEnvironmentVariableW`.
> 
> In short, I'm not entirely sure how you'd look for likely suspects. I guess if they set a `Command` var with lower case ASCII? But that's only a problem if the child process absolutely requires ASCII upper case.

Thank you for this summary. In this context I'm much less worried about the potential breakage here.

Based on the second half of my comment [here](https://github.com/rust-lang/rust/pull/85270#issuecomment-854257301) I'm gonna go ahead and start an FCP for this.

@rfcbot merge

File: library/std/src/sys/windows/process.rs (hunk 6 above)
Comment (m-ou-se) [reply library/std/src/sys/windows/process.rs:46]: (@CDirkx Unrelated to this PR: Note that this happens even on Unix, because OsString is not zero terminated. So every time one has to be passed to the OS, it gets copied to a new allocation to add a zero terminator, which is deallocated again right afterwards. Unfortunately, 'OsString' means 'can represent every string that the OS accepts', not 'a string that the OS can directly use'. We can't change this unless we start a 'std 2.0' or something.)

File: library/std/src/sys/windows/process.rs (hunk 3 above)
Comment (m-ou-se) [reply library/std/src/sys/windows/process.rs]: Can you add this conclusion, why CompareStringOrdinal is used, in a comment in the code?

Comment (ChrisDenton) [reply library/std/src/sys/windows/process.rs]: @m-ou-se I've [added a comment](https://github.com/rust-lang/rust/pull/85270/commits/a200c01e4fb32179af1cbbc47a1e5ea7a6394e9a) (based on CDirkx's) to briefly explain how comparing environment keys works and thus why `CompareStringOrdinal` is used.

**Comment by ChrisDenton** [comment]
Hm, it seems that test is (unintentionally) testing specifically for Rust upper casing key names. Which is wrong even without this PR because the intent is presumably to test user interaction. I've opened an issue, linked above, that details the problem.

**Comment by ChrisDenton** [comment]
Ok, the failure is addressed here: https://github.com/rust-lang/cargo/pull/9646

If I'm understanding right the problem was that cargo:

1. It always spawns itself
2. In doing so it sets an environment variable, using `env`, which caused Rust to upper case all environment variables
3. It had Windows specific tests that assumed lower cased environment variables would always be converted to upper case (which was previously true for cargo itself because of 1 and 2).

If this PR goes ahead I think the release notes should emphasize that, on Windows, using `env` with `Command` would previously cause all environment variables to be ASCII-uppercased.
